modules/frontend/traceid_sharder.go

```diff
@@ -66,35 +63,38 @@ func (s asyncTraceSharder) RoundTrip(pipelineRequest pipeline.Request) (pipeline
 	}
 
 	return pipeline.NewAsyncSharderFunc(ctx, int(concurrentShards), len(reqs), func(i int) pipeline.Request {
-		pipelineReq := pipelineRequest.CloneFromHTTPRequest(reqs[i])
+		pipelineReq := reqs[i]
 		return pipelineReq
 	}, s.next), nil
 }
 
 // buildShardedRequests returns a slice of requests sharded on the precalculated
 // block boundaries
-func (s *asyncTraceSharder) buildShardedRequests(ctx context.Context, parent *http.Request) ([]*http.Request, error) {
+func (s *asyncTraceSharder) buildShardedRequests(parent pipeline.Request) ([]pipeline.Request, error) {
 	userID, err := user.ExtractOrgID(parent.Context())
 	if err != nil {
 		return nil, err
 	}
 
-	reqs := make([]*http.Request, s.cfg.QueryShards)
+	reqs := make([]pipeline.Request, s.cfg.QueryShards)
 	params := map[string]string{}
 	// build sharded block queries
 	for i := 0; i < len(s.blockBoundaries); i++ {
-		reqs[i] = parent.Clone(ctx)
-		if i == 0 {
-			// ingester query
-			params[querier.QueryModeKey] = querier.QueryModeIngesters
-		} else {
-			// block queries
-			params[querier.BlockStartKey] = hex.EncodeToString(s.blockBoundaries[i-1])
-			params[querier.BlockEndKey] = hex.EncodeToString(s.blockBoundaries[i])
-			params[querier.QueryModeKey] = querier.QueryModeBlocks
-		}
-		reqs[i] = api.BuildQueryRequest(reqs[i], params)
-		prepareRequestForQueriers(reqs[i], userID)
+		i := i // save the loop variable locally to make sure the closure grabs the correct var.
+		pipelineR, _ := cloneChildRequest(parent, userID, func(r *http.Request) (*http.Request, error) {
+			if i == 0 {
+				// ingester query
+				params[querier.QueryModeKey] = querier.QueryModeIngesters
```

Comment:
Maybe we could take this out and start the loop on 1

Reply:
sure. i was trying to change the original code as little as possible, but agree it's a nicer structure. change made!